File: 백준/Gold/20208. 진우의 민트초코우유/진우의 민트초코우유.py
import sys
input = sys.stdin.readline
dx = [-1, 1, 0, 0]
dy = [0, 0, 1, -1]
def dfs(x, y, hp, mCnt):
    global max_milk
    if abs(x - home[0])+abs(y - home[1]) <= hp:
        max_milk = max(mCnt, max_milk)

    for i in range(len(milk)):
        if not visited[i]:
            mx = milk[i][0]
            my = milk[i][1]
            dis = abs(mx - x) + abs(my - y)
            if dis <= hp:
                visited[i] = 1
                dfs(mx, my, hp - dis + h, mCnt + 1)
                visited[i] = 0

n, m, h = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
home = [0, 0]
max_milk = 0
milk = []
for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            home = [i, j]
        elif graph[i][j] == 2:
            milk.append([i,j])

visited = [0] * len(milk)
dfs(home[0], home[1], m, 0)
print(max_milk)

File: 백준/Gold/20208. 진우의 민트초코우유/진우의 민트초코우유.java
import java.awt.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
import java.util.List;

import static java.lang.Long.sum;
import static java.lang.Math.abs;
import static java.lang.Math.max;

public class Main {

    static int N,M,H;
    static int sx, sy;
    static int graph[][];
    static List<Point> milk;
    static int ans;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());
        milk = new ArrayList<>();
        sx= 0; sy= 0;
        ans = 0;
        graph = new int[N][N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
                if (graph[i][j] == 1) {
                    sx = i; sy = j;
                }
                else if (graph[i][j] == 2) {
                    milk.add(new Point(i, j));
                }
            }
        }
        dfs(sx, sy, M, 0);
        System.out.println(ans);

    }
    static void dfs(int x, int y, int hp, int mCnt){
        for (int i = 0; i < milk.size(); i++) {
            int nx = milk.get(i).x;
            int ny = milk.get(i).y;
            if (graph[nx][ny] == 2) {
                int dist = Math.abs(x - nx) + Math.abs(y - ny);
                if (dist <= hp) {
                    graph[nx][ny] = 0;
                    dfs(nx, ny, hp - dist + H, mCnt + 1);
                    graph[nx][ny] = 2;
                }
            }
        }
        int hdist = Math.abs(x - sx) + Math.abs(y - sy);
        if (hdist <= hp) {
            ans = Math.max(mCnt, ans);
        }

    }
}